algorithms/rainbow/main.py

```diff
@@ -0,0 +1,177 @@
+                    help='Display screen (testing only)')
+
+if __name__ == '__main__':
+    # Setup
+    args = parser.parse_args()
+    print(' ' * 26 + 'Options')
+    for k, v in vars(args).items():
+        print(' ' * 26 + k + ': ' + str(v))
+    np.random.seed(args.seed)
+    torch.manual_seed(np.random.randint(1, 10000))
+    if torch.cuda.is_available() and not args.disable_cuda:
+        args.device = torch.device('cuda')
+        torch.cuda.manual_seed(np.random.randint(1, 10000))
+        # Disable non deterministic ops (not sure if critical but better safe than sorry)
+        torch.backends.cudnn.enabled = False
+    else:
+        args.device = torch.device('cpu')
+
+    # Simple ISO 8601 timestamped logger
+    def log(s):
+        print('[' + str(datetime.now().strftime('%Y-%m-%dT%H:%M:%S')) + '] ' + s)
+
+    # Environment selection
+    if args.game == 'ai2thor':
+        env = MultipleStepsEnv(AI2ThorEnv(
+            config_file='config_files/rainbow_example.json'),
+                               args.history_length, args.device)
+        args.resolution = env.env.config['resolution']
+    else:
+        env = Env(args)
+        env.train()
+        args.resolution = (84, 84)
+    action_space = env.action_space
+
+    # Agent
+    dqn = Agent(args, env)
+    mem = ReplayMemory(args, args.memory_capacity)
+    priority_weight_increase = (1 - args.priority_weight) / (args.T_max - args.learn_start)
+
+    # Construct validation memory
+    val_mem = ReplayMemory(args, args.evaluation_size)
+    T, done = 0, True
+    for T in range(args.evaluation_size):
+        if done:
+            state, done = env.reset(), False
+        next_state, _, done, _ = env.step(env.action_space.sample())
+        val_mem.append(state, None, None, done)
+        state = next_state
+
+    if args.evaluate:
+        dqn.eval()  # Set DQN (online network) to evaluation mode
+        avg_reward, avg_Q = test(env, T, args, dqn, val_mem, evaluate=True)
+        print('Avg. reward: ' + str(avg_reward) + ' | Avg. Q: ' + str(avg_Q))
+    else:
+        # Training loop
+        dqn.train()
+        T, done = 0, True
+
+        while T < args.T_max:
+            if T % 200 == 0:
+                print("step {}".format(T))
+
+            if done:
+                state, done = env.reset(), False
+            if T % args.replay_frequency == 0:
+                dqn.reset_noise()  # Draw a new set of noisy epsilons
+
+            action = dqn.act(state)  # Choose an action greedily (with noisy weights)
+            next_state, reward, done, _ = env.step(action)  # Step
+            if args.reward_clip > 0:
+                reward = max(min(reward, args.reward_clip), -args.reward_clip)  # Clip rewards
+            mem.append(state, action, reward, done)  # Append transition to memory
+            T += 1
+
+            if T % args.log_interval == 0:
+                log('T = ' + str(T) + ' / ' + str(args.T_max))
```

Comment:
interesting function and we might prefer to log to a file in the future so might be good to update the function but maybe not for this PR